### PR TITLE
Composition polish and bug fixes

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Closes https://github.com/crossplane/crossplane/pull/1445
Closes https://github.com/crossplane/crossplane/pull/1450
Closes https://github.com/crossplane/crossplane/pull/1449

This PR fixes potential release blocker bugs in our new composition controllers and includes a bunch of "polish" refactoring.

Fixes for known issues include:

* Composite resources can be deleted even if their composition no longer exists.
* Composite resources derive a reclaim policy and secret namespace from their composition.
* Composite resources become `Ready` when all composed resources are `Ready`.
* Composed resource secrets don't need to conflict with their composite resource secret.
* Resource requirements don't bind to compositions until they are `Ready`.
* Resource requirements can successfully unbind from their composite resources.
* Resource requirements respect the reclaim policy of their resource compositions.
* An `InfrastructureDefinition` cannot be deleted while its `InfrastructurePublication` exists.

Unfixed known issues that I do not consider release blockers include:

* Updating a resource requirement does not update its bound composite resource.
* An `InfrastructurePublication`is not deleted when its `InfrastructureDefinition` is.
* Composed resources could be (statically) bound by a a good-old-fashioned resource claim.

"Polish" includes:

* Controllers watch the resources they own or relate to, where possible.
* Refactoring such that dependencies are injectable for when we add tests.
* All controllers emit debug logs and events.
* Our partial validation schemas in `ccrd` are stored in their unmarshalled form.
* The `ccrd` package is tested.
* Definition kinds are now "Established" (like a CRD), not "Ready".

### How has this code been tested?

I've added a few unit tests, but otherwise all testing is manual. I'd like to merge without tests (given these controllers already have no coverage) and focus on adding tests tomorrow. See comments below for details on how I've tested this manually.

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation] and [examples].
- [ ] Reported all new error conditions into the log or as an event, as
  appropriate.

For more about what we believe makes a pull request complete, see our
[definition of done].

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[definition of done]: https://github.com/crossplane/crossplane/tree/master/design/one-pager-definition-of-done.md

### CODE REVIEW

**Comment by negz** [comment]
To manually test this I used two files:

`id.yaml`:

```yaml
---
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: InfrastructurePublication
metadata:
  name: mysqls.db.example.org
spec:
  infrastructureDefinitionRef:
    name: mysqls.db.example.org
---
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: InfrastructureDefinition
metadata:
  name: mysqls.db.example.org
spec:
  connectionSecretKeys:
    - username
    - password
    - endpoint
  crdSpecTemplate:
    group: db.example.org
    version: v1alpha1
    names:
      kind: MySQL
      listKind: MySQLList
      plural: mysqls
      singular: mysql
    validation:
      openAPIV3Schema:
        type: object
        properties:
          spec:
            type: object
            properties:
              engineVersion:
                enum:
                  - "5.6"
                  - "5.7"
                type: string
              storageGB:
                type: integer
---
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Composition
metadata:
  name: mysqlcomposition
  labels:
    connectivity: private
spec:
  writeConnectionSecretsToNamespace: crossplane-system
  reclaimPolicy: Delete
  from:
    apiVersion: db.example.org/v1alpha1
    kind: MySQL
  to:
    - base:
        apiVersion: azure.crossplane.io/v1alpha3
        kind: ResourceGroup
        spec:
          location: West US 2
          reclaimPolicy: Delete
          providerRef:
            name: example
    - base:
        apiVersion: database.azure.crossplane.io/v1beta1
        kind: MySQLServer
        spec:
          forProvider:
            administratorLogin: myadmin
            resourceGroupNameSelector:
              matchControllerRef: true
            location: West US 2
            sslEnforcement: Disabled
            version: "5.6"
            sku:
              tier: GeneralPurpose
              capacity: 2
              family: Gen5
            storageProfile:
              storageMB: 8192
              backupRetentionDays: 7
              geoRedundantBackup: Disabled
          writeConnectionSecretToRef:
            namespace: crossplane-system
          providerRef:
            name: example
          reclaimPolicy: Delete
      patches:
      - fromFieldPath: "metadata.uid"
        toFieldPath: "spec.writeConnectionSecretToRef.name"
        # Note the new transform.
        transforms:
        - type: string
          string:
            fmt: "%s-sql"
      - fromFieldPath: "spec.engineVersion"
        toFieldPath: "spec.forProvider.version"
      - fromFieldPath: "spec.storageGB"
        toFieldPath: "spec.forProvider.storageMB"
        transforms:
          - type: math
            math:
              multiply: 1024
      connectionDetails:
        - fromConnectionSecretKey: username
        - fromConnectionSecretKey: password
        - fromConnectionSecretKey: endpoint
```

`sql.yaml`:

```yaml
---
apiVersion: db.example.org/v1alpha1
kind: MySQLRequirement
metadata:
  namespace: default
  name: example-sql
spec:
  engineVersion: "5.7"
  storageGB: 2
  compositionSelector:
    matchLabels:
      connectivity: private
  writeConnectionSecretToRef:
    name: example-sql
```

</details>

**Comment by negz** [comment]
Applying `id.yaml` creates the `InfrastructureDefinition`, `InfrastructurePublication`, and `Composition`. Events and debug logs show the CRDs are rendered and created, and the controllers started.

```
$ kubectl apply -f id.yaml 
infrastructurepublication.apiextensions.crossplane.io/mysqls.db.example.org created
infrastructuredefinition.apiextensions.crossplane.io/mysqls.db.example.org created
composition.apiextensions.crossplane.io/mysqlcomposition created

$ kubectl get crd|grep example.org
mysqlrequirements.db.example.org                                   2020-04-28T12:05:13Z
mysqls.db.example.org                                              2020-04-28T12:05:13Z
```

```
$ kubectl describe infrastructuredefinition
Name:         mysqls.db.example.org
Namespace:    
Labels:       <none>
Annotations:  API Version:  apiextensions.crossplane.io/v1alpha1
Kind:         InfrastructureDefinition
Metadata:
  Creation Timestamp:  2020-04-28T12:05:13Z
  Finalizers:
    finalizer.apiextensions.crossplane.io
    published.apiextensions.crossplane.io
  Generation:        1
  Resource Version:  374571
  Self Link:         /apis/apiextensions.crossplane.io/v1alpha1/infrastructuredefinitions/mysqls.db.example.org
  UID:               219e99d3-eac4-44ab-9311-f46f96a05e20
Spec:
  Connection Secret Keys:
    username
    password
    endpoint
  Crd Spec Template:
    Group:  db.example.org
    Names:
      Kind:       MySQL
      List Kind:  MySQLList
      Plural:     mysqls
      Singular:   mysql
    Validation:
      openAPIV3Schema:
        Properties:
          Spec:
            Properties:
              Engine Version:
                Enum:
                  5.6
                  5.7
                Type:  string
              Storage GB:
                Type:  integer
            Type:      object
        Type:          object
    Version:           v1alpha1
Status:
  Conditions:
    Last Transition Time:  2020-04-28T12:05:13Z
    Reason:                Created CRD and started controller
    Status:                True
    Type:                  Established
    Last Transition Time:  2020-04-28T12:05:13Z
    Reason:                Successfully reconciled resource
    Status:                True
    Type:                  Synced
Events:
  Type    Reason                         Age                From                                                                Message
  ----    ------                         ----               ----                                                                -------
  Normal  ApplyInfrastructureDefinition  42s                apiextensions/infrastructuredefinition.apiextensions.crossplane.io  waiting for CustomResourceDefinition to be established
  Normal  ApplyInfrastructureDefinition  42s (x2 over 42s)  apiextensions/infrastructuredefinition.apiextensions.crossplane.io  Applied CustomResourceDefinition and (re)started composite controller
```

```
$ kubectl describe infrastructurepublication
Name:         mysqls.db.example.org
Namespace:    
Labels:       <none>
Annotations:  API Version:  apiextensions.crossplane.io/v1alpha1
Kind:         InfrastructurePublication
Metadata:
  Creation Timestamp:  2020-04-28T12:05:13Z
  Finalizers:
    published.apiextensions.crossplane.io
  Generation:        1
  Resource Version:  374578
  Self Link:         /apis/apiextensions.crossplane.io/v1alpha1/infrastructurepublications/mysqls.db.example.org
  UID:               f4e0f124-6b75-4731-b55f-714dd880ff5b
Spec:
  Infrastructure Definition Ref:
    Name:  mysqls.db.example.org
Status:
  Conditions:
    Last Transition Time:  2020-04-28T12:05:13Z
    Reason:                Successfully reconciled resource
    Status:                True
    Type:                  Synced
    Last Transition Time:  2020-04-28T12:05:13Z
    Reason:                Created CRD and started controller
    Status:                True
    Type:                  Established
Events:
  Type     Reason                          Age                From                                                                 Message
  ----     ------                          ----               ----                                                                 -------
  Warning  GetInfrastructureDefinition     76s (x2 over 76s)  apiextensions/infrastructurepublication.apiextensions.crossplane.io  InfrastructureDefinition.apiextensions.crossplane.io "mysqls.db.example.org" not found
  Warning  ApplyInfrastructurePublication  76s                apiextensions/infrastructurepublication.apiextensions.crossplane.io  cannot update object: Operation cannot be fulfilled on infrastructuredefinitions.apiextensions.crossplane.io "mysqls.db.example.org": the object has been modified; please apply your changes to the latest version and try again
  Normal   ApplyInfrastructurePublication  76s (x2 over 76s)  apiextensions/infrastructurepublication.apiextensions.crossplane.io  waiting for CustomResourceDefinition to be established
  Normal   GetInfrastructureDefinition     73s (x6 over 76s)  apiextensions/infrastructurepublication.apiextensions.crossplane.io  Got published InfrastructureDefinition
  Normal   RenderCustomResourceDefinition  73s (x6 over 76s)  apiextensions/infrastructurepublication.apiextensions.crossplane.io  Rendered CustomResourceDefinition
  Normal   ApplyInfrastructurePublication  73s (x3 over 76s)  apiextensions/infrastructurepublication.apiextensions.crossplane.io  Applied CustomResourceDefinition and (re)started requirement controller
```

**Comment by negz** [comment]
Applying `sql.yaml` creates the `MySQLRequirement`, which in turn creates a `MySQL`. Note that the `MySQL` and thus the `MySQLRequirement` are not `Ready` because the resources they compose are not yet `Ready`.

```
$ kubectl apply -f sql.yaml 
mysqlrequirement.db.example.org/example-sql created
```

```
$ kubectl describe mysqlrequirementName:         example-sql
Namespace:    default
Labels:       <none>
Annotations:  API Version:  db.example.org/v1alpha1
Kind:         MySQLRequirement
Metadata:
  Creation Timestamp:  2020-04-28T12:11:48Z
  Generation:          2
  Resource Version:    375132
  Self Link:           /apis/db.example.org/v1alpha1/namespaces/default/mysqlrequirements/example-sql
  UID:                 22705c82-7ff9-4c19-8e90-2d1d519f5f71
Spec:
  Composition Selector:
    Match Labels:
      Connectivity:  private
  Engine Version:    5.7
  Resource Ref:
    Name:      default-example-sql-dkrgp
  Storage GB:  2
  Write Connection Secret To Ref:
    Name:  example-sql
Status:
  Conditions:
    Last Transition Time:  2020-04-28T12:11:48Z
    Reason:                Resource requirement is waiting for composite resource to become Ready
    Status:                False
    Type:                  Ready
    Last Transition Time:  2020-04-28T12:11:48Z
    Reason:                Successfully reconciled resource
    Status:                True
    Type:                  Synced
Events:
  Type    Reason                      Age                From                               Message
  ----    ------                      ----               ----                               -------
  Normal  ConfigureCompositeResource  69s                requirement/mysqls.db.example.org  Successfully configured composite resource
  Normal  BindCompositeResource       69s (x4 over 69s)  requirement/mysqls.db.example.org  Composite resource is not yet ready
```

```
$ kubectl describe mysql
Name:         default-example-sql-dkrgp
Namespace:    
Labels:       <none>
Annotations:  <none>
API Version:  db.example.org/v1alpha1
Kind:         MySQL
Metadata:
  Creation Timestamp:  2020-04-28T12:11:48Z
  Generate Name:       default-example-sql-
  Generation:          4
  Resource Version:    375145
  Self Link:           /apis/db.example.org/v1alpha1/mysqls/default-example-sql-dkrgp
  UID:                 ba9e9399-4a5a-4e0c-b9e8-5c95bab43687
Spec:
  Composition Ref:
    Name:  mysqlcomposition
  Composition Selector:
    Match Labels:
      Connectivity:  private
  Engine Version:    5.7
  Reclaim Policy:    Delete
  Requirement Ref:
    Name:       example-sql
    Namespace:  default
  Resource Refs:
    API Version:  azure.crossplane.io/v1alpha3
    Kind:         ResourceGroup
    Name:         default-example-sql-dkrgp-n7fqv
    UID:          6714539a-cde2-4cc6-8ea5-535e47873623
    API Version:  database.azure.crossplane.io/v1beta1
    Kind:         MySQLServer
    Name:         default-example-sql-dkrgp-tq994
    UID:          bc4836f0-a4e2-4b9d-a8c1-64cff1874181
  Storage GB:     2
  Write Connection Secret To Ref:
    Name:       ba9e9399-4a5a-4e0c-b9e8-5c95bab43687
    Namespace:  crossplane-system
Status:
  Conditions:
    Last Transition Time:  2020-04-28T12:11:48Z
    Reason:                Resource is being created
    Status:                False
    Type:                  Ready
    Last Transition Time:  2020-04-28T12:11:48Z
    Reason:                Successfully reconciled resource
    Status:                True
    Type:                  Synced
Events:
  Type    Reason                   Age                From                             Message
  ----    ------                   ----               ----                             -------
  Normal  SelectComposition        53s (x3 over 83s)  composite/mysqls.db.example.org  Successfully selected composition
  Normal  PublishConnectionSecret  53s (x6 over 83s)  composite/mysqls.db.example.org  Successfully published connection details
  Normal  ComposeResources         53s (x3 over 83s)  composite/mysqls.db.example.org  Successfully composed resources
```

```
$ kubectl describe resourcegroup|grep -B3 Ready 
    Last Transition Time:  2020-04-28T12:11:53Z
    Reason:                Resource is available for use
    Status:                True
    Type:                  Ready

$ kubectl describe mysqlserver|grep -B3 Ready 
    Last Transition Time:  2020-04-28T12:11:55Z
    Reason:                Resource is being created
    Status:                False
    Type:                  Ready
```

**Comment by negz** [comment]
Once the composed resources become `Ready` the composite and requirement become `Ready` too:

```
$ kubectl describe mysqlserver|grep -B3 Ready 
    Last Transition Time:  2020-04-28T12:17:44Z
    Reason:                Resource is available for use
    Status:                True
    Type:                  Ready

$ kubectl describe resourcegroup|grep -B3 Ready 
    Last Transition Time:  2020-04-28T12:11:53Z
    Reason:                Resource is available for use
    Status:                True
    Type:                  Ready
```

```
$ kubectl describe mysql
Name:         default-example-sql-dkrgp
Namespace:    
Labels:       <none>
Annotations:  <none>
API Version:  db.example.org/v1alpha1
Kind:         MySQL
Metadata:
  Creation Timestamp:  2020-04-28T12:11:48Z
  Generate Name:       default-example-sql-
  Generation:          4
  Resource Version:    375724
  Self Link:           /apis/db.example.org/v1alpha1/mysqls/default-example-sql-dkrgp
  UID:                 ba9e9399-4a5a-4e0c-b9e8-5c95bab43687
Spec:
  Composition Ref:
    Name:  mysqlcomposition
  Composition Selector:
    Match Labels:
      Connectivity:  private
  Engine Version:    5.7
  Reclaim Policy:    Delete
  Requirement Ref:
    Name:       example-sql
    Namespace:  default
  Resource Refs:
    API Version:  azure.crossplane.io/v1alpha3
    Kind:         ResourceGroup
    Name:         default-example-sql-dkrgp-n7fqv
    UID:          6714539a-cde2-4cc6-8ea5-535e47873623
    API Version:  database.azure.crossplane.io/v1beta1
    Kind:         MySQLServer
    Name:         default-example-sql-dkrgp-tq994
    UID:          bc4836f0-a4e2-4b9d-a8c1-64cff1874181
  Storage GB:     2
  Write Connection Secret To Ref:
    Name:       ba9e9399-4a5a-4e0c-b9e8-5c95bab43687
    Namespace:  crossplane-system
Status:
  Conditions:
    Last Transition Time:  2020-04-28T12:18:19Z
    Reason:                Resource is available for use
    Status:                True
    Type:                  Ready
    Last Transition Time:  2020-04-28T12:11:48Z
    Reason:                Successfully reconciled resource
    Status:                True
    Type:                  Synced
Events:
  Type    Reason                   Age                     From                             Message
  ----    ------                   ----                    ----                             -------
  Normal  PublishConnectionSecret  6m26s (x12 over 9m57s)  composite/mysqls.db.example.org  Successfully published connection details
  Normal  ComposeResources         6m26s (x6 over 9m57s)   composite/mysqls.db.example.org  Successfully composed resources
  Normal  SelectComposition        4m26s (x8 over 9m57s)   composite/mysqls.db.example.org  Successfully selected composition
```

```
$ kubectl describe mysqlrequirement
Name:         example-sql
Namespace:    default
Labels:       <none>
Annotations:  API Version:  db.example.org/v1alpha1
Kind:         MySQLRequirement
Metadata:
  Creation Timestamp:  2020-04-28T12:11:48Z
  Finalizers:
    finalizer.apiextensions.crossplane.io
  Generation:        2
  Resource Version:  375730
  Self Link:         /apis/db.example.org/v1alpha1/namespaces/default/mysqlrequirements/example-sql
  UID:               22705c82-7ff9-4c19-8e90-2d1d519f5f71
Spec:
  Composition Selector:
    Match Labels:
      Connectivity:  private
  Engine Version:    5.7
  Resource Ref:
    Name:      default-example-sql-dkrgp
  Storage GB:  2
  Write Connection Secret To Ref:
    Name:  example-sql
Status:
  Conditions:
    Last Transition Time:  2020-04-28T12:18:19Z
    Reason:                Resource is available for use
    Status:                True
    Type:                  Ready
    Last Transition Time:  2020-04-28T12:11:48Z
    Reason:                Successfully reconciled resource
    Status:                True
    Type:                  Synced
Events:
  Type    Reason                      Age                    From                               Message
  ----    ------                      ----                   ----                               -------
  Normal  ConfigureCompositeResource  10m                    requirement/mysqls.db.example.org  Successfully configured composite resource
  Normal  BindCompositeResource       10m (x4 over 10m)      requirement/mysqls.db.example.org  Composite resource is not yet ready
  Normal  BindCompositeResource       4m11s (x2 over 4m11s)  requirement/mysqls.db.example.org  Successfully bound composite resource
  Normal  PropagateConnectionSecret   4m11s (x2 over 4m11s)  requirement/mysqls.db.example.org  Successfully propagated connection details from composite resource
```

**Comment by negz** [comment]
All three connection secrets were written:

```
# The requirement
$ kubectl get -o json mysqlrequirement example-sql | jq .spec.writeConnectionSecretToRef.name
"example-sql"

# The composite resource
$ kubectl get -o json mysql default-example-sql-dkrgp | jq .spec.writeConnectionSecretToRef.name
"ba9e9399-4a5a-4e0c-b9e8-5c95bab43687"

# The composed resource
$ kubectl get -o json mysqlserver default-example-sql-dkrgp-tq994 | jq .spec.writeConnectionSecretToRef.name
"ba9e9399-4a5a-4e0c-b9e8-5c95bab43687-sql"
```

And all three contain the same keys:

```
# The requirement
$ kubectl get -o json -n default secret example-sql|jq .data
{
  "endpoint": "REDACTED",
  "password": "REDACTED",
  "username": "bXlhZG1pbkBkZWZhdWx0LWV4YW1wbGUtc3FsLWRrcmdwLXRxOTk0"
}

# The composite resource
$ kubectl get -o json -n crossplane-system secret ba9e9399-4a5a-4e0c-b9e8-5c95bab43687|jq .data
{
  "endpoint": "REDACTED",
  "password": "REDACTED",
  "username": "bXlhZG1pbkBkZWZhdWx0LWV4YW1wbGUtc3FsLWRrcmdwLXRxOTk0"
}

# The composed resource
$ kubectl get -o json -n crossplane-system secret ba9e9399-4a5a-4e0c-b9e8-5c95bab43687-sql|jq .data
{
  "endpoint": "REDACTED",
  "password": "REDACTED",
  "username": "bXlhZG1pbkBkZWZhdWx0LWV4YW1wbGUtc3FsLWRrcmdwLXRxOTk0"
}
```

**Comment by negz** [comment]
Deleting the requirement deletes the composite resource and the composed resources:

```
$ kubectl delete -f sql.yaml 
mysqlrequirement.db.example.org "example-sql" deleted

$ kubectl get mysql
No resources found in default namespace.

# Resource groups take about 100 forevers to delete.
$ kubectl describe resourcegroup|grep elete
  Reclaim Policy:  Delete
  Normal  DeletedExternalResource  12s (x3 over 13s)  managed/resourcegroup.azure.crossplane.io  Successfully requested deletion of external resource

$ kubectl describe mysqlserver|grep elete
No resources found in default namespace.
```

**Comment by negz** [comment]
Finally, if you delete the `InfrastructureDefinition`, `InfrastructurePublication`, and the `Composite` all at once then they and all of the resources they define will be removed:

```
$ kubectl apply -f id.yaml 
infrastructurepublication.apiextensions.crossplane.io/mysqls.db.example.org created
infrastructuredefinition.apiextensions.crossplane.io/mysqls.db.example.org created
composition.apiextensions.crossplane.io/mysqlcomposition created

$ kubectl apply -f sql.yaml 
mysqlrequirement.db.example.org/example-sql created

$ kubectl get mysqlrequirement,mysql,mysqlserver,resourcegroup
NAME                                          AGE
mysqlrequirement.db.example.org/example-sql   3s

NAME                                             AGE
mysql.db.example.org/default-example-sql-mmgjh   3s

NAME                                                                       STATUS   STATE   CLASS   VERSION   AGE
mysqlserver.database.azure.crossplane.io/default-example-sql-mmgjh-ctbq8                                      3s

NAME                                                                AGE
resourcegroup.azure.crossplane.io/default-example-sql-mmgjh-jk5xd   3s

$ kubectl delete -f id.yaml 
infrastructurepublication.apiextensions.crossplane.io "mysqls.db.example.org" deleted
infrastructuredefinition.apiextensions.crossplane.io "mysqls.db.example.org" deleted
composition.apiextensions.crossplane.io "mysqlcomposition" deleted

$ sleep 120 # Resource groups are sloooow.
$ kubectl get mysqlrequirement,mysql,mysqlserver,resourcegroup
Error from server (NotFound): Unable to list "db.example.org/v1alpha1, Resource=mysqlrequirements": the server could not find the requested resource (get mysqlrequirements.db.example.org)
Error from server (NotFound): Unable to list "db.example.org/v1alpha1, Resource=mysqls": the server could not find the requested resource (get mysqls.db.example.org)
```

File: pkg/controller/apiextensions/publication/watch.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/composite"
+)
+
+type adder interface {
+	Add(item interface{})
+}
+
+// EnqueueRequestForRequirement enqueues a reconcile.Request for the
+// NamespacedName of a RequirementReferencer's RequirementReference.
+type EnqueueRequestForRequirement struct{}
```
Comment (muvaf) [inline pkg/controller/apiextensions/publication/watch.go:36]: Why is this necessary?

File: pkg/controller/apiextensions/requirement/api.go [hunk 2]
```diff
@@ -102,8 +111,7 @@ func (a *APIBinder) Bind(ctx context.Context, rq resource.Requirement, cp resour
 // removing the composite resource's requirement reference, and if the composite
 // resource's reclaim policy is "Delete", deleting it.
 func (a *APIBinder) Unbind(ctx context.Context, _ resource.Requirement, cp resource.Composite) error {
-	// TODO(negz): Does this work for unstructured resources?
-	cp.SetRequirementReference(nil)
```
Comment (muvaf) [inline pkg/controller/apiextensions/requirement/api.go:106]: I was expecting this to work. Why doesn't it?

Comment (negz) [reply pkg/controller/apiextensions/requirement/api.go:106]: It sets the ref to the JSON `null`, which is not a valid value according to the schema validation. I believe `*unstructured.Unstructured` has special casing here that interprets `nil` as "delete this field" - we might want to do the same in `pkg/fieldpath`.

File: pkg/controller/apiextensions/publication/watch.go (hunk 1 above)
Comment (negz) [reply pkg/controller/apiextensions/publication/watch.go:36]: It's analogous to the existing `resource.EnqueueRequestForClaim` from crossplane-runtime. It allows the requirement controller to watch composite resources of the kind that it is interested in (e.g. the `MySQLRequirement` controller can watch for `MySQL`). This allows the requirement controller to avoid speculatively requeuing reconciles when it's waiting for the composite resource to do something; it can instead just return `Requeue: false` and wait for a watch event.

We can't use the upstream `EnqueueRequestForOwner{}` here, because the (namespaced) requirement is not the owner of the (cluster scoped) composite.

File: pkg/controller/apiextensions/publication/reconciler.go [hunk 3]
```diff
@@ -164,159 +293,172 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
...
+		r.record.Event(p, event.Normal(reasonDeletePub, "Stopped requirement controller and deleted CustomResourceDefinition"))
 		p.Status.SetConditions(runtimev1alpha1.ReconcileSuccess())
-		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, p), errUpdateInfraPubStatus)
+		return reconcile.Result{RequeueAfter: tinyWait}, errors.Wrap(r.client.Status().Update(ctx, p), errUpdateStatus)
 	}
 
-	// We will do some operations now that we will have to cleanup, so, we need
-	// a finalizer to have the chance.
-	if err := r.AddFinalizer(ctx, p); err != nil {
+	// Note that we add a finalizer to the InfrastructureDefinition that we
```
[the remark this answers is not in the document]
Comment (negz) [reply pkg/controller/apiextensions/publication/reconciler.go:407]: I had a similar thought about "publication" being a field/flag of the definition rather than a distinct CR. It might be worth exploring these approaches post-release.

File: pkg/controller/apiextensions/publication/reconciler.go [hunk 4]
```diff
@@ -164,159 +293,172 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
...
-		// Controller should be stopped only after all instances are gone so that
-		// deletion logic of the instances are processed by the controller.
+		// Ensure all the custom resources we defined are gone before stopping
+		// the controller we started to reconcile them. This ensures the
+		// controller has a chance to execute its cleanup logic, if any.
 		if len(l.Items) > 0 {
-			// TODO(negz): For some reason DeleteAllOf does not work here,
-			// despite working just fine in the definition controller. Possibly
-			// due to requirements being namespaced rather than cluster scoped?
+			// TODO(negz): DeleteAllOf does not work here, despite working in
```
[the remark this answers is not in the document]
Comment (negz) [reply pkg/controller/apiextensions/publication/reconciler.go:366]: I'm 99% sure I did. I want to try again just in case, but will do so in a future PR.

**Comment by negz** [comment]
Okay, I addressed @muvaf's feedback and tested this again. I'm going to merge so we can move forward with adding tests.
